prow/github/client_test.go

```diff
@@ -795,49 +795,88 @@ func TestUnassignIssue(t *testing.T) {
 }
 
 func TestReadPaginatedResults(t *testing.T) {
-	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if r.Method != http.MethodGet {
-			t.Errorf("Bad method: %s", r.Method)
-		}
-		if r.URL.Path == "/label/foo" {
-			objects := []Label{{Name: "foo"}}
-			b, err := json.Marshal(objects)
-			if err != nil {
-				t.Fatalf("Didn't expect error: %v", err)
+	type response struct {
+		labels []Label
+		next   string
+	}
+	cases := []struct {
+		name        string
+		baseSuffix  string
+		initialPath string
+		responses   map[string]response
+	}{
+		{
+			name:        "regular pagination",
+			initialPath: "/label/foo",
+			responses: map[string]response{
+				"/label/foo": {
+					labels: []Label{{Name: "foo"}},
+					next:   `<blorp>; rel="first", <https://%s/label/bar>; rel="next"`,
+				},
+				"/label/bar": {
+					labels: []Label{{Name: "bar"}},
+				},
+			},
+		},
+		{
+			name:        "pagination with /api/v3 base suffix",
+			initialPath: "/label/foo",
+			baseSuffix:  "/api/v3",
+			responses: map[string]response{
+				"/api/v3/label/foo": {
+					labels: []Label{{Name: "foo"}},
+					next:   `<blorp>; rel="first", <https://%s/api/v3/label/bar>; rel="next"`,
+				},
+				"/api/v3/label/bar": {
+					labels: []Label{{Name: "bar"}},
+				},
+			},
+		},
+	}
+	for _, tc := range cases {
+		ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.Method != http.MethodGet {
+				t.Errorf("Bad method: %s", r.Method)
 			}
-			w.Header().Set("Link", fmt.Sprintf(`<blorp>; rel="first", <https://%s/label/bar>; rel="next"`, r.Host))
-			fmt.Fprint(w, string(b))
-		} else if r.URL.Path == "/label/bar" {
-			objects := []Label{{Name: "bar"}}
-			b, err := json.Marshal(objects)
-			if err != nil {
-				t.Fatalf("Didn't expect error: %v", err)
+			if response, ok := tc.responses[r.URL.Path]; ok {
+				b, err := json.Marshal(response.labels)
+				if err != nil {
+					t.Fatalf("Didn't expect error: %v", err)
+				}
+				if response.next != "" {
+					w.Header().Set("Link", fmt.Sprintf(response.next, r.Host))
+				}
+				fmt.Fprint(w, string(b))
+			} else {
+				t.Errorf("Bad request path: %s", r.URL.Path)
 			}
-			fmt.Fprint(w, string(b))
+		}))
+		defer ts.Close()
+
+		c := getClient(ts.URL)
+		c.bases[0] = c.bases[0] + tc.baseSuffix
+		var labels []Label
+		err := c.readPaginatedResults(
+			tc.initialPath,
+			"",
+			func() interface{} {
+				return &[]Label{}
+			},
+			func(obj interface{}) {
+				labels = append(labels, *(obj.(*[]Label))...)
+			},
+		)
+		if err != nil {
+			t.Errorf("%s: didn't expect error: %v", tc.name, err)
 		} else {
-			t.Errorf("Bad request path: %s", r.URL.Path)
+			var expectedLabels []Label
+			for _, response := range tc.responses {
```

Comment:
I am not really a fan of a test that determines its outcome at runtime (and might have bugs there) rather than having it hardcoded. I am btw also completely fine with just moving the five relevant lines here into a new function and then just testing that new functions, which is probably easier than adjusting this test.

Reply:
Makes sense. I added `expectedLabels` to the test case.

I would keep this test as table test if it's okay. I think it's easier to see this way what the actual differences between both supported variants are. It's also easier to extend in the future compared to the old test if necessary.